feat(astro): experimental astro:env (#10974)
withastro/astro: .changeset/poor-berries-occur.md

---
"astro": minor
---

> Adds experimental support for the `astro:env` API. It lets you define a type-safe schema for your environment variables and specify their availability on the server or client.

Author: Updated the changeset, ptal!

> Adds experimental support for the `astro:env` API. It lets you define a type-safe schema for your environments variables, and where they should be available (on the server or the client).

Reviewer: ```suggestion
Adds experimental support for the `astro:env` API. It lets you define a type-safe schema for your environment variables and specify their availability on the server or client.
```

Start by declaring a schema in your Astro config in `experimental.env.schema`. Use the `envField` help to make your life easier:

```js
// astro.config.mjs
import { defineConfig, envField } from "astro/config"

export default defineConfig({
    experimental: {
        env: {
            schema: {
                PUBLIC_API_URL: envField.string({ context: "client", access: "public", optional: true }),
                PUBLIC_PORT: envField.number({ context: "server", access: "public", default: 4321 }),

>                 API_SECRET: envField.string({ context: "server", access: "secret" }),

Reviewer 2: Re-reading this, since `PUBLIC_` already hinted what the `access` would be, maybe we should remove `access` altogether from the expose API, and we can auto set it internally?

Author: I quite like forcing people to be super explicit but I don't mind inferring it as well. I'd like other opinions before refactoring

Reviewer 3: I would also prefer expressiveness over "magic"

Reviewer 2: Alright then. Let's see how this goes.

            }
        }
    }
})
```

There are currently 3 data types supported: strings, numbers and booleans.

There are also 3 kinds of variables:

> - **Public client variables**: Those variables end up in your final client and server bundle and can be accessed both client and server through the `astro:env/client` module:

Reviewer: ```suggestion
- **Public client variables**: These variables are included in both the client and server bundles and can be accessed from either via the astro:env/client module:
```

    ```js
    import { PUBLIC_API_URL } from "astro:env/client"
    ```

> - **Public server variables**: Those variables end up in your final server bundle and can be accessed on the server through the `astro:env/server` module:

Reviewer: ```suggestion
- **Public server variables**: These variables are included only in the server bundle and can be accessed on the server via the astro:env/server module:
```

    ```js
    import { PUBLIC_PORT } from "astro:env/server"
    ```

> - **Secret server variables**: Those variables are not part of your final bundle and can be accessed on the server through the `astro:env/server` module:

Reviewer: ```suggestion
- **Secret server variables**: These variables are not included in the final bundle and can only be accessed on the server via the astro:env/server module:
```

```astro
---
import { PUBLIC_APP_ID } from "astro:env/client"
import { PUBLIC_API_URL, getSecret } from "astro:env/server"
const API_TOKEN = getSecret("API_TOKEN")

const data = await fetch(`${PUBLIC_API_URL}/users`, {
	method: "POST",
	headers: {
		"Content-Type": "application/json",
		"Authorization": `Bearer ${API_TOKEN}`
	},
	body: JSON.stringify({ appId: PUBLIC_APP_ID })
})
---
```

To define the data type and properties of your environment variables, declare a schema in your Astro config in `experimental.env.schema`. The `envField` helper allows you define your variable as a string, number, or boolean and pass properties in an object:

```js
// astro.config.mjs
import { defineConfig, envField } from "astro/config"

export default defineConfig({
    experimental: {
        env: {
            schema: {
                PUBLIC_API_URL: envField.string({ context: "client", access: "public", optional: true }),
                PUBLIC_PORT: envField.number({ context: "server", access: "public", default: 4321 }),
                API_SECRET: envField.string({ context: "server", access: "secret" }),
            }
        }
    }
})
```

> There are currently 3 data types supported: strings, numbers and booleans.
> 
> There are three kinds of variables, determined by the combination of `context` (`client` or `server`) and `access` (`private` or `public`) settings defined in your [`env.schema`](#experimentalenvschema):

Reviewer 4: ```suggestion
There are three kinds of environment variables, determined by the combination of `context` (`client` or `server`) and `access` (`private` or `public`) settings defined in your [`env.schema`](#experimentalenvschema):
```
I think for the changeset, we don't need to repeat the three kinds of data types, since we said them above. Also, I like explicitly "environment variables" instead of just variables here.
